packages/amplify_authenticator/lib/src/widgets/form.dart

```diff
@@ -262,35 +232,66 @@ class SignUpForm extends AuthenticatorForm {
   /// {@macro amplify_authenticator.sign_up_form}
   SignUpForm({
     Key? key,
-  }) : this.custom(
-          fields: [],
+  })  : useRunTimeFields = true,
+        super._(
           key: key,
+          fields: [
+            SignUpFormField.username(),
+            SignUpFormField.password(),
+            SignUpFormField.passwordConfirmation(),
+          ],
+          actions: const [
+            SignUpButton(),
+          ],
         );
 
   /// A custom Sign Up form.
-  SignUpForm.custom({
+  const SignUpForm.custom({
     Key? key,
     required List<SignUpFormField> fields,
-  }) : super._(
+  })  : useRunTimeFields = false,
+        super._(
           key: key,
-          requiredFields: [
-            SignUpFormField.username(),
-            SignUpFormField.password(),
-            SignUpFormField.passwordConfirmation(),
-          ],
-          customFields: fields,
+          fields: fields,
           actions: const [
             SignUpButton(),
           ],
         );
 
+  /// determines if run time fields should be used
+  final bool useRunTimeFields;
+
   @override
   _SignUpFormState createState() => _SignUpFormState();
+  @override
+  void debugFillProperties(DiagnosticPropertiesBuilder properties) {
+    super.debugFillProperties(properties);
+    properties
+        .add(DiagnosticsProperty<bool>('useRunTimeFields', useRunTimeFields));
+  }
 }
 
 class _SignUpFormState extends AuthenticatorFormState<SignUpForm> {
   _SignUpFormState() : super();
 
+  @override
+  List<AuthenticatorFormField> get allFields {
+    if (!widget.useRunTimeFields) {
+      return widget.fields;
+    }
+
+    // combine fields and sort on priority
+    final fields = [
+      ...widget.fields,
+      ...runtimeFields(context),
+    ]..sort((a, b) {
```

Comment:
This will sort the custom fields and `sort` is not stable. So, fields could theoretically get jumbled up every time you load the form.

Reply:
I updated it to use mergeSort from flutter/foundation, which is stable.

Reply:
Noice

Reply:
The other issue, though, is that the order of the custom fields may be different when including runtime fields since it's not sorted then. Is that expected?

Reply:
There are two options for creating a SignUpForm. The default constructor `SignUpForm()` and the named constructor `SignUpForm.custom()`. `includeDefaultFields` is always true when the default constructor is used, and is always false when the named constructor is used. When the default constructor is used, the consumer has no control over what fields are displayed. The required username/password fields are combined with the fields from the config, and should be sorted. When the named constructor is used, the consumer has full control over what fields are displayed, and they should be displayed in the order they are passed in (they should not be sorted).

So I don't think it is an issue that fields are sorted when `includeDefaultFields` is true and are not sorted when `includeDefaultFields` is false. It is intentional.

It would potentially be an issue if `SignUpForm.custom()` accepted `includeDefaultFields` as an argument. But `includeDefaultFields` is only used internally. It could actually be made private. I will push an update to make it private.

Reply:
Ah, yeah I got a bit confused. Sounds good.